drivers/store/redis/store.go

```diff
@@ -3,42 +3,81 @@ package redis
 import (
 	"context"
 	"fmt"
+	"strings"
 	"time"
 
 	libredis "github.com/go-redis/redis/v8"
 	"github.com/pkg/errors"
-
 	"github.com/ulule/limiter/v3"
 	"github.com/ulule/limiter/v3/drivers/store/common"
 )
 
+const (
+	luaIncrScript = `
+local key = KEYS[1]
+local count = tonumber(ARGV[1])
+local ttl = tonumber(ARGV[2])
+local ret = redis.call("incrby", key, ARGV[1])
+if ret == count then
+	if ttl > 0 then
+		redis.call("pexpire", key, ARGV[2])
+	end
+	return {ret, ttl}
+end
+ttl = redis.call("pttl", key)
+return {ret, ttl}
+`
+	luaPeekScript = `
+local key = KEYS[1]
+local v = redis.call("get", key)
+if v == false then
+	return {0, 0}
+end
+local ttl = redis.call("pttl", key)
+return {tonumber(v), ttl}
+`
+)
+
 // Client is an interface thats allows to use a redis cluster or a redis single client seamlessly.
 type Client interface {
 	Get(ctx context.Context, key string) *libredis.StringCmd
 	Set(ctx context.Context, key string, value interface{}, expiration time.Duration) *libredis.StatusCmd
 	Watch(ctx context.Context, handler func(*libredis.Tx) error, keys ...string) error
 	Del(ctx context.Context, keys ...string) *libredis.IntCmd
 	SetNX(ctx context.Context, key string, value interface{}, expiration time.Duration) *libredis.BoolCmd
-	Eval(ctx context.Context, script string, keys []string, args ...interface{}) *libredis.Cmd
+	EvalSha(ctx context.Context, sha string, keys []string, args ...interface{}) *libredis.Cmd
+	ScriptLoad(ctx context.Context, script string) *libredis.StringCmd
 }
 
 // Store is the redis store.
 type Store struct {
 	// Prefix used for the key.
 	Prefix string
+	// deprecated, this option make no sense when all operations were atomic
 	// MaxRetry is the maximum number of retry under race conditions.
 	MaxRetry int
 	// client used to communicate with redis server.
 	client Client
+	// luaIncrSHA is the SHA of increase and expire key script
+	luaIncrSHA string
+	// luaPeekSHA is the SHA of peek and expire key script
+	luaPeekSHA string
 }
 
 // NewStore returns an instance of redis store with defaults.
 func NewStore(client Client) (limiter.Store, error) {
-	return NewStoreWithOptions(client, limiter.StoreOptions{
+	store, err := NewStoreWithOptions(client, limiter.StoreOptions{
 		Prefix:          limiter.DefaultPrefix,
 		CleanUpInterval: limiter.DefaultCleanUpInterval,
 		MaxRetry:        limiter.DefaultMaxRetry,
 	})
+	if err != nil {
+		return nil, err
+	}
+	if err := store.(*Store).preloadLuaScripts(context.Background()); err != nil {
```

Comment:
Could you do the `preloadLuaScripts` inside `NewStoreWithOptions`, this way you don't have to cast the `store`, and it's available for any constructor.